### PR TITLE
[1.24] Backported automatic registration

### PR DESCRIPTION
**Pull request by jirihnidek**
* Card ID: 4741
* Original PR: #2407
  - Original commit (part of it): a4b70b31ca204deb533e822a94986edf4a7a6b4a
* It is possible to perform automatic registration on AWS, Azure
  and GCP using subscription-manager-1.24
* Modified code on few places to make it compatible with Python2
* Fixed three unit tests

### CODE REVIEW

File: etc-conf/rhsm.conf [hunk 1]
```diff
@@ -94,6 +94,12 @@ autoAttachInterval = 1440
 splay = 1
 # If set to 1, rhsmcertd will not execute.
 disable = 0
+# Set to 1, when rhsmcertd will try to do automatic registration.
```
Comment (cnsnyder) [inline etc-conf/rhsm.conf:97]: I'd suggest we change this line to "Set to 1 when rhsmcertd should attempt automatic registration"

File: etc-conf/rhsmcertd.completion.sh [hunk 2]
```diff
@@ -11,7 +11,7 @@ _rhsmcertd()
 	first="${COMP_WORDS[1]}"
 	cur="${COMP_WORDS[COMP_CWORD]}"
 	prev="${COMP_WORDS[COMP_CWORD-1]}"
-	opts="-h --help -c --cert-check-interval --cert-interval -d --debug --heal-interval -i --auto-attach-interval -n --now -s --no-splay"
+	opts="-h --help -c --cert-check-interval --cert-interval -d --debug --heal-interval -i --auto-attach-interval -n --now -s --no-splay  -a --auto-registration -r --auto-registration-interval"
```
Comment (cnsnyder) [inline etc-conf/rhsmcertd.completion.sh:14]: Do we need the two spaces in between "--no-splay" and "-a" or is one ok?

File: man/rhsm.conf.5 [hunk 3]
```diff
@@ -247,7 +247,17 @@ were set to 3 minutes, the initial cert check would begin somewhere between 2 mi
 .PP
 disable
 .RS 4
-set to 1 to disable rhsmcertd operation entirely.
+Set to 1 to disable rhsmcertd operation entirely.
+.RE
+.PP
+auto_registration
+.RS 4
+Set to 1 to enabled automatic registration. Automatic registration can only work on virtual machines running in the public cloud. Currently three public cloud providers are supported: AWS, Azure and GCP. In order for rhsmcertd to perform automatic registration, please link your "Cloud ID" from your cloud provider to your "RHSM Organization ID" using https://cloud.redhat.com.
```
Comment (cnsnyder) [inline man/rhsm.conf.5:255]: enabled -> enable

File: src/rhsmlib/services/register.py [hunk 4]
```diff
@@ -128,5 +130,8 @@ def validate_options(self, options):
             elif options['environment']:
                 raise exceptions.ValidationError(_("Error: Activation keys do not allow environments to be"
                     " specified."))
+        elif options.get('jwt_token') is not None:
+            # TODO: add more checks here
```
Comment (cnsnyder) [inline src/rhsmlib/services/register.py:134]: Is this TODO for another PR?

**Review by cnsnyder — CHANGES_REQUESTED** [review]
Aside from a few places I have suggestions on the wording of a few comments, this looks good to me! Please take a look at the quick suggestions and we'll go from there! Thank you very much!

File: src/rhsmlib/services/register.py (hunk 4 above)
Comment (jirihnidek) [reply src/rhsmlib/services/register.py:134]: I don't know what was my intention with this TODO. We can delete.

**Comment by jirihnidek** [comment]
@cnsnyder updated PR. When these fix of grammar mistakes will be approved, then we should fix it for main branch too.

**Pull request merged**